[Patrick Tan] iP

pattan9955 opened the pull request:
### PepePro

> “Your mind is for having ideas, not holding them.” – David Allen (source)

PepePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/pattan9955/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

pattan9955 committed:
Implemented skeletal version of Duke with echo and exit functionality.

Added a Response class to deal with string formatting for responses.

pattan9955 committed:
Implemented adding of items to list.
Implemented listing of stored items.

pattan9955 committed:
1) Implemented marking of tasks as completed.
2) Added a Task class to represent an individual Task.
3) Added a Tasklist class to handle the storing of Tasks in a list.
4) Edited Response.java to handle the printing of formatted strings as well.

pattan9955 committed:
Added tracking for ToDo, Event and Deadline tasks.
1) Implemented counting of total tasks in list on addition of task.
2) Added individual classes for each task type to handle individual behaviour.

pattan9955 committed:
Performed semi-automated testing of text UI
1) Tested basic functionality (startup text, event, todo, deadline, list, done, bye)
2) Fixed a bug where created events' descriptions would appear twice in the UI, stray print statement in code.

pattan9955 committed:
Added error handling for multiple edge cases.
1) Handled cases where the input to "done" is not a number or out of range.
2) Handled cases where "todo" is called without a valid description.
3) Handled cases where "deadline" is called without a valid description or due date.
4) Handled cases where "event" is called without a valid description or event date.

pattan9955 committed:
Implemented delete functionality for tasks.
1) Fixed error handling for "done" when no argument was given.

pattan9955 committed:
Implemented bug fix for parsing
1) Split by " /at " and " /by " for events and deadlines respectively instead of "/at" and "/by".

pattan9955 committed:
Add persistent storage functionality to Duke
The Duke application does not have any persistent storage ability, which means that tasks created during the lifespan of a single run of the application persists only until the user calls the "bye" command.

Creating a text file that acts as a form of storage for Duke will help to rectify this.

Let's implement the persistent storage in a class called PersistentStorage that deals with both loading tasks from and saving tasks to said text file.

pattan9955 committed:
Add DateTime parsing for Deadlines and Events
Due dates and event date times are stored as Strings which does not allow for searches by date or time.

Changing the internal representation of Deadline and Event objects to use Java DateTimes will allow us to rectify this.

Let's change the constructors for Deadline and Event objects to accept a Java DateTime instead, which will be stored as an instance variable.

pattan9955 committed:
Merge branch 'branch-Level-7'

pattan9955 committed:
Merge branch 'branch-Level-8'

pattan9955 committed:
Update Persistent Storage to be compatible with use of Java DateTimes.

pattan9955 committed:
Refactor code to improve abstraction and code design

Parsing, reading and execution of commands were all done in the main body of Duke.java which led to poor readability.

Abstracting each function(parsing, reading, UI, persistence etc) improves the quality of the code and the readability.

Let's abstract each function into its own class, and include additional classes for individual commands.

pattan9955 committed:
Organize classes into packages for easy navigation

Having all the classes in the main /java directory makes the application disorganized and hard to navigate.

Dividing the classes into folders based on their logical groupings (e.g. commands, tasks) makes the project easier to navigate and organize.

Let's group the Command related classes (e.g. DeadlineCommand.java) and the Task related classes (e.g. ToDo.java) into their own packages.

pattan9955 committed:
Add JUnit tests

There is a need for validation of the various Duke classes.

In order to achieve this, JUnit tests that focus on testing the Duke commands and tasks classes are required.

Let's do so by adding JUnit tests that test the aforementioned classes.

pattan9955 committed:
Release v0.1 of Duke

Release Duke in executable .jar format.

pattan9955 committed:
Add JavaDocs for all classes in Duke

pattan9955 committed:
Tweak code to comply with coding standard

Some classes currently do not fully comply with the coding standard, with examples being wildcard imports in some classes.

Tweaking the code to comply with the coding standard improves the readability of the code.

Let's ensure the codebase follows this coding standard by making necessary edits to code style.

pattan9955 committed:
Implement find functionality

Duke did not have a search functionality which made it hard for users to find previously made tasks.

The introduction of this command will help improve user friendliness of Duke.

Let's implement find functionality as a Command type and update the UI and Tasklistss classes to be able to accommodate this functionality.

pattan9955 committed:
Merge branch 'branch-A-JavaDoc'

pattan9955 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Tasklist.java
#	src/main/java/duke/UI.java
#	src/main/java/duke/commands/ByeCommand.java
#	src/main/java/duke/commands/Command.java
#	src/main/java/duke/commands/DeadlineCommand.java
#	src/main/java/duke/commands/DeleteCommand.java
#	src/main/java/duke/commands/DoneCommand.java
#	src/main/java/duke/commands/EventCommand.java
#	src/main/java/duke/commands/TodoCommand.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Task.java
#	src/main/java/duke/tasks/ToDo.java

pattan9955 committed:
Fix for minor typo in JavaDoc

pattan9955 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/UI.java

pattan9955 committed:
Fix FindCommand import in Parser.java

limzk126 commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,43 @@
+public class ByeCommand extends Command {
+
+    public ByeCommand(){}
+
```
Missing comment for public constructor?

limzk126 commented on src/main/java/duke/commands/ListAllCommand.java:
```diff
@@ -0,0 +1,25 @@
+public class ListAllCommand extends Command {
+
+    public ListAllCommand(){}
+
```
Same as mentioned earlier.

limzk126 commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,87 @@
+
+    /** A boolean representing if the task is complete */
+    private Boolean completed;
+
```
Would isCompleted sound more boolean?

limzk126 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    public void run() {
+        ui.showStartMsg();
+        Boolean exit = false;
+
```
Would isExit sound more boolean?

limzk126 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,163 @@
+
+        // Check if fullCommand is "list"
+        else if (fullCommand.equals("list")) {
+            return new ListAllCommand();
```
Should the else-if statements be on the same line as the closing parenthesis '}'?

pattan9955 replied:
Indeed they should be! I have made the changes, thank you for pointing these out! :)

limzk126 reviewed (commented):
Good java coding standards overall, probably just a few nits to fix. 👍

tianyue58 commented on src/main/java/duke/PersistentStorage.java:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Constructor for a PersistentStorage
+     *
+     * @param filepath The filepath to the text file used for
```
Should punctuation be added behind the parameter description? 🤔

tianyue58 commented on src/main/java/duke/Tasklist.java:
```diff
@@ -0,0 +1,124 @@
+
+/**
+ * Class that handles the storing of Tasks during the app's runtime
+  */
```
Should the subsequent * (such as * on L9) be aligned with the first one? 🤔

tianyue58 commented on src/main/java/duke/Tasklist.java:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * Returns all Tasks whose descriptions contain any of the provided search terms.
+     *
+     * @param searchTerms A String describing the search terms to check the Tasks against
```
Should punctuation be added behind each parameter description?

tianyue58 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,303 @@
+        /**
+         * A constructor for a Response object.
+         *
+         * @param content A String containing the content of the Response
```
Should punctuation be added behind each parameter description?

tianyue58 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,303 @@
+         * @param task The Task that has been completed
+         * @return A Response object detailing the completion of the Task
+         */
+        public static Response completed(Task task) {
```
Is a method name such as completeTask() more appropriate? 🤔

pattan9955 replied:
Indeed it is! Thank you for pointing all these out! :)

tianyue58 reviewed (commented):
LGTM. Just some small things to fix! :)

pattan9955 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

pattan9955 committed:
Merge branch 'branch-A-Gradle'

pattan9955 committed:
Add CheckStyle checking functionality and fix coding standard violations
There is a need to ensure code adheres to a coding standard in order to maintain its readability.

Adding CheckStyle functionality that can be used together with Gradle can help to ensure a fixed coding standard throughout the project.

Let's add the configuration files for the CheckStyle functionality and fix any coding standard violations found by CheckStyle.

pattan9955 committed:
Merge branch 'branch-A-CheckStyle'

pattan9955 committed:
Fix more coding standard violations in test classes

YeluriKetan commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,297 @@
+            Response response = new Response(msg);
+            System.out.println(response);
+            return response;
+        }
```
I can see why you named these methods the way you did and have `showXXX` methods for functionality, but do you think it would be better to name the methods using verbs like `createRemoveResponse`?

YeluriKetan commented on src/test/java/DukeCommandsTest.java:
```diff
@@ -0,0 +1,38 @@
+import duke.commands.DeadlineCommand;
+
+public class DukeCommandsTest {
+    @Test
+    public void byeCommandTest() {
+        Command c = new ByeCommand();
+        assertEquals(true, c.isExit());
+    }
+
+    @Test
+    public void deadlineCommandTest() {
+        UI ui = new UI(new Scanner(System.in));
+        Tasklist taskList = new Tasklist();
+        PersistentStorage storage = new PersistentStorage("./test.txt");
+
+        DeadlineCommand c = new DeadlineCommand("test description",
+                LocalDateTime.of(2012, 2, 2, 16, 30));
+
+        c.executeCommand(taskList, ui, storage);
+        String expected = "1. [D][ ] test description (by: Feb 02 2012 16:30)";
+
+        assertEquals(expected, taskList.toString());
+
+
+    }
```
Could perhaps name the test methods in a more descriptive manner, in the form of `featureUnderTest_testScenario_expectedBehavior()`?

YeluriKetan reviewed (commented):
LGTM, very few nits to fix that are hard to find. JavaDocs is very well done and shows that you have put in time and effort. The code is neatly formatted and easy to understand because of step-by-step explanation using comments within a method. Good luck!!!

pattan9955 committed:
Add GUI for Duke
Interaction with the Duke application is done completely through the command line interface.

Adding a GUI will help to improve the overall user experience and make it more approachable to new users.

To that end, let's implement the GUI using JavaFX and remove the existing command line interface.

pattan9955 committed:
Merge branch 'branch-Level-10'

pattan9955 committed:
Add assertions to main codebase

A good number of functions and methods in Duke utilize some form of user input which is prone to errors.
This is particularly true for methods that deal with parsing of user data.

Ensuring certain assumptions about the code holds will help with ensuring the correctness of the code now
and in future.

As a step towards such an assurance, let's add assertions to the codebase, in particular the Parser and Tasklist
classes, in particular for methods that utilize user input. Doing so will help ensure the correctness of the code.

pattan9955 committed:
Fix codebase's code quality

There are several methods in Duke that do not obey best practices
for coding quality, with several lengthy methods, particularly in
Parser.

Refactoring the relevant portions makes for more readable and
better documented code, that improves the ease of future
development.

To this end, let's refactor the code base by abstracting repeated
portions of code into separate methods, and shortening the lengthy
"parse" method using switch statements and enums. This will help
improve the quality of the code overall.

pattan9955 committed:
Merge pull request #2 from pattan9955/branch-A-CodeQuality

Fix codebase's code quality

pattan9955 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/util/Tasklist.java

pattan9955 committed:
Merge pull request #3 from pattan9955/branch-A-Assertions

Branch a assertions

pattan9955 committed:
Add support for DoAfter tasks

Support already exists for a variety of task types, including
ToDos, Deadlines and Events, with the ability to add datetime
information for relevant task types.

Adding support for DoAfter tasks provides end-users with more
flexibility in task management when using the app.

To this end, let's implement support for a "doafter" command which
allows users to create a DoAfter task by
* Updating the Parser to parse "doafter" commands
* Updating PersistentStorage to read/write DoAfter tasks to file
* Adding a new DoAfterCommand type

pattan9955 committed:
Merge branch 'branch-B-DoAfterTasks'

pattan9955 committed:
Improve GUI of Duke

The GUI of Duke needs to add value to the program, with one such way
being added readability.

Adding visual elements to the GUI that help with the readability would
help with the overall user experience.

To this end, let's demarcate the different dialog boxes used (by Duke vs
the user) through the use of different colours, and clip the profile
pictures into a circle for aesthetics.

pattan9955 committed:
Merge branch 'branch-A-BetterGui'

pattan9955 committed:
Implement bug fix for loading tasks from file

Task strings stored in the persistent file were not validated before
loading into the Duke application, which led to the application
crashing if a malformed persistent storage file was provided.

Validating the task strings stored in the persistent file improves the
resilience of the application.

To this end, let's
* implement a validation function in PersistentStorage
that validates task strings
* implement checking in the Duke GUI to inform the user should the
persistence file not be successfully loaded

pattan9955 committed:
Update user guide and UI screenshot for Duke

pattan9955 committed:
Set theme jekyll-theme-midnight

pattan9955 committed:
Edit colour scheme of GUI

pattan9955 committed:
Update UI screenshot

pattan9955 committed:
Clean up codebase and neaten code

The Parser in Duke was too long and led to unsightly code. In
addition, validation was done by a separate function that led to
unnecessary code.

To fix this, more abstractions in the parsing of the commands in
the Parser class could lead to cleaner code.

To this end, let us
* abstract each switch case in the Parser class into its own
function

A bug where resizing the Duke GUI would lead to unexpected
window behaviours was also fixed. The GUI is now non-resizable.
